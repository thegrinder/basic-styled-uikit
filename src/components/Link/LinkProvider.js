import { createProvider } from '../../helpers';

const LinkProvider = createProvider('link', {});

export default LinkProvider;
