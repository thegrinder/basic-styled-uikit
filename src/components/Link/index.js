export { default as Link } from './Link';
export { default as LinkProvider } from './LinkProvider';
