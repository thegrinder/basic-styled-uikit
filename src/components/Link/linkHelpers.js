import { getStyle } from '../../helpers/utils';

const getLinkType = ({ theme, linkType }, state) => theme.uiKit.link.linkTypes[linkType][state];
const getLinkTypeStyle = state => getStyle(getLinkType, state);

export const linkNormalStyle = getLinkTypeStyle('normal');
export const linkHoverStyle = getLinkTypeStyle('hover');
