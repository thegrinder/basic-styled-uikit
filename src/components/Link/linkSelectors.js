import { getLink, getMode } from '../../theme/themeSelectors';

export const getLinkTypeStyle = (theme, linktype, state) => {
  const mode = getMode(theme);
  const typeStyle = getLink(theme).linkTypes[mode][linktype];
  if (!typeStyle) {
    throw new Error(`There is no ${linktype} link type in the <Link /> theme`);
  }
  if (!typeStyle[state]) {
    throw new Error(
      `There is no ${state} state for ${linktype} link type in <Link /> theme`
    );
  }
  return typeStyle[state];
};

export const getLinkCommonStyle = theme => getLink(theme).common;
