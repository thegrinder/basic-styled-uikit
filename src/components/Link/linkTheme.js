import { darken, lighten } from 'polished';
import colors from '../../theme/colors';

const linkTheme = {
  linkTypes: {
    default: {
      light: {
        normal: {
          color: colors.primary600,
        },
        hover: {
          color: darken(0.1, colors.primary600),
        },
      },
      dark: {
        normal: {
          color: colors.primary400,
        },
        hover: {
          color: lighten(0.1, colors.primary400),
        },
      },
    },
    muted: {
      light: {
        normal: {
          color: colors.neutral600,
        },
        hover: {
          color: darken(0.1, colors.neutral600),
        },
      },
      dark: {
        normal: {
          color: colors.neutral400,
        },
        hover: {
          color: lighten(0.1, colors.neutral400),
        },
      },
    },
  },
  common: {
    fontFamily: 'Lato, Arial, Helvetica, sans-serif',
  },
};

export default linkTheme;
