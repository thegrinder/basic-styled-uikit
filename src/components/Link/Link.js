import styled from 'styled-components';
import { oneOf } from 'prop-types';
import { linkNormalStyle, linkHoverStyle } from './linkHelpers';

const Link = styled.a`
  cursor: pointer;
  touch-action: manipulation;
  text-decoration: none;
  font-family: inherit;
  ${linkNormalStyle}
  &:hover {
    ${linkHoverStyle}
    outline: none;
    text-decoration: underline;
  }
`;

Link.propTypes = {
  linkType: oneOf(['default', 'muted']),
};

Link.defaultProps = {
  linkType: 'default',
};

export default Link;
