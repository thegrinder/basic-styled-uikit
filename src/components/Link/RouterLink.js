import React from 'react';
import styled from 'styled-components';
import { oneOf } from 'prop-types';
import { Link as ReactRouterLink } from 'react-router-dom';
import sharedStyles from '../../helpers/sharedStyles';
import { linkNormalStyle, linkHoverStyle } from './linkHelpers';

const Link = styled(({ linkType, ...rest }) => <ReactRouterLink {...rest} />)`
  ${sharedStyles}
  cursor: pointer;
  touch-action: manipulation;
  text-decoration: none;
  font-family: inherit;
  ${linkNormalStyle}
  &:hover {
    ${linkHoverStyle}
    outline: none;
    text-decoration: underline;
  }
`;

Link.propTypes = {
  linkType: oneOf(['default', 'muted']),
};

Link.defaultProps = {
  linkType: 'default',
};

export default Link;
