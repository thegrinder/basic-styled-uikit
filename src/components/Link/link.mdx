---
name: Link
menu: Components
route: /components/link
---

import { Playground, Props } from 'docz';
import { TypographyProvider } from '../Typography';
import LinkProvider from './LinkProvider';
import Link from './Link';

## **LinkProvider**

`<LinkProvider />` is responsible for passing down the theme to the `<Link />` component. It expects the theme to be of certain structure.



### **Example**

<Playground>
  {() => {
    const linkTheme = {
      linkTypes: {
        default: {
          normal: {
            color: '#1e88e5',
          },
          hover: {
            color: '#166dba',
          },
        },
        muted: {
          normal: {
            color: '#868e96',
          },
          hover: {
            color: '#6c747d',
          },
        },
      },
      common: {
        fontWeight: 'bold',
        fontFamily: `'Inter var', sans-serif`,
      },
    };

    const Example = () => (
      <TypographyProvider>
        <LinkProvider theme={linkTheme}>
          <>
            <Link className="margin-bottom">default link</Link>
            <Link linkType="muted">muted link</Link>
          </>
        </LinkProvider>
      </TypographyProvider>
    );

    return <Example />;
  }}
</Playground>

---



## **Link**

Link component grabs the theme given to `<LinkProvider />` and applies the style defined there based on the given `linkType` prop value.

<Props of={Link} />
