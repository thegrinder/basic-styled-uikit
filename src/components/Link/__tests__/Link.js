import React from 'react';
import { render } from '../../../test-utils';
import Link from '../Link';

const children = 'children';

const renderComponent = (props = {}) =>
  render(<Link {...props}>{children}</Link>);

describe('<Link />', () => {
  it('should render correctly with default props and its children', () => {
    const { queryByText } = renderComponent();
    expect(queryByText(children)).toBeTruthy();
  });

  it('should render correctly with custom props', () => {
    const { queryByText } = renderComponent({
      sizing: 'xs',
      linktype: 'muted',
    });
    expect(queryByText(children)).toBeTruthy();
  });

  it('should render <a> tag by default', () => {
    const { queryByText } = renderComponent();
    expect(queryByText(children).tagName).toEqual('A');
  });

  it('should render render <button> tag', () => {
    const { queryByText } = renderComponent({ as: 'button' });
    expect(queryByText(children).tagName).toEqual('BUTTON');
  });
});
