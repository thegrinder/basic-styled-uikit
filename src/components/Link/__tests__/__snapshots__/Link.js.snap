// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> should render correctly with custom props 1`] = `
<a
  class="StyledLink-sc-1s1vukk-0 lazzqj"
>
  children
</a>
`;

exports[`<Link /> should render correctly with default props and its children 1`] = `
<a
  class="StyledLink-sc-1s1vukk-0 ivYnxS"
>
  children
</a>
`;

exports[`<Link /> should render render <button> tag 1`] = `
<button
  class="StyledLink-sc-1s1vukk-0 ivYnxS"
>
  children
</button>
`;
